### PR TITLE
add es-modules support

### PR DESCRIPTION
**Pull request by MikeDevice**
This is useful when library is used in applications fully based on ECMAScript Modules. In this case all configuration files for `east` is written on ES Modules. Even if not they can require other ES Modules modules inside.

```js
// .eastrc

import config from '../config';

export default {
  dir: path.resolve(config.path)
};
```

Now `east` can not work inside such applications. It throws a lot of different errors. So I was thinking how to fix it. I came up with the following solutions.

1. Write all `east` configuration files in CommonJS. That looks like as a best solution, but in real application it is too complicated, because configuration files can require a lot of modules from app's core, like config or something else, so this solution makes us to write some files using ES Modules and CommonJS, or just use Babel for server files what is not a good idea. So this solution makes us to change our source code because `east` can't work with `import/export`.

2. Add ES Modules support to `east`, what I did in this pull-request. The main idea of my solution is to replace all dynamic `require` statements to a new util `requireModule` which makes `require()` or `import()` depending on `esModules` flag.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -22,11 +22,11 @@
...
-    "makeCodeCoverageSummaryReport": "npm run testSpecified test -- --reporter dot --coverage --coverage-report text-summary test | sed -re 's/(Statements|Branches|Functions|Lines)/\\1 code coverage/g'",
+    "testWithNode": "for file in `find test -name *.js | grep -v es-modules`; do node \"$file\" || exit 1; done;",
+    "test": "NODE_OPTIONS='--experimental-modules --experimental-specifier-resolution=node --no-warnings' npm run testSpecified test",
+    "makeTestSnapshots": "NODE_OPTIONS='--experimental-modules --experimental-specifier-resolution=node --no-warnings' TAP_SNAPSHOT=1 npm run testSpecified test",
+    "makeCodeCoverageDetailReport": "npm run testSpecified test -- --grep es-modules --invert --reporter dot --coverage --coverage-report html --no-browser test && echo \"*** Code coverage report done, checkout ./coverage dir\"",
+    "makeCodeCoverageSummaryReport": "npm run testSpecified test -- --grep es-modules --invert --reporter dot --coverage --coverage-report text-summary test | sed -re 's/(Statements|Branches|Functions|Lines)/\\1 code coverage/g'",
```
Comment (MikeDevice) [inline package.json:29]: There is a bug with coverage and `NODE_OPTIONS='--experimental-modules --experimental-specifier-resolution=node'`, so I decided to skip `es-modules` test for coverage

File: lib/utils/requireModule/.eslintrc [hunk 2]
```diff
@@ -0,0 +1,3 @@
+{
+	"parser": "babel-eslint"
```
Comment (MikeDevice) [inline lib/utils/requireModule/.eslintrc:2]: This `parser` option is for only one purpose: correct parse the `import` statements.

File: lib/utils/requireModule/importModule.js [hunk 3]
```diff
@@ -0,0 +1,6 @@
+'use strict';
+
+module.exports = (path) => {
+	return import(path)
```
Comment (MikeDevice) [inline lib/utils/requireModule/importModule.js:4]: > This parser option is for only one purpose: correct parsing for import statements.

Like this. Btw, this is the only one `import` statement

**Comment by okv** [comment]
Idea of this PR is interesting, I'll try to take a closer look shortly.

**Comment by okv** [comment]
I'm getting my head around using es and commonjs modules together in nodejs, so it takes some time. I have a bunch of suggestions to this PR but the main idea looks good to me. I'll try to put my suggestions here soon.

**Comment by okv** [comment]
Here are my main suggestions:

* rebase this PR on [feature/v2](https://github.com/okv/east/pull/50) which drops old nodejs versions support and contains updated code base
* turn `utils.requireModule(path, esModules)` into `utils.loadModule({path, esModules})`. Here is an idea to use `load` verb which internally will do `requireMudule` or `importModule`. Single object function signature is commonly used at other methods and seems reasonable here. I also think that `loadModule` should always return object (`module.default || module` seems to
be ok but should be always converted to object e.g. using `_.extend`)
* add info to changelog and readme about this feature (config, migrations, adapter and plugins could be provided as es modules with this flag)

and couple of small things:

* please try not to use nested chains of promises (like it currently does at `_createAdapter`, `_registerPlugins`) - use flat chain instead (like at other methods)
* add migrate, rollback snapshot tests with esmodules migrations e.g. as 
test/02-snapshot/bin/east/commands/migrate/withEsModuleMigration.js, 
test/02-snapshot/bin/east/commands/rollback/withEsModuleMigration.js
* treat `esModules` as one of the `params` on commands level
* treat `esMoudles` as one of the `params` at `migrator.configure` instead of
constructor
* get rid of babel-eslint (if it's possible) e.g. by adding file with import to eslintignore

**Comment by okv** [comment]
@MikeDevice what do you think about that?

**Comment by MikeDevice** [comment]
I like it. Sorry for the lack of response. I was a little busy with other projects.

**Comment by MikeDevice** [comment]
I'll try to fix this PR as soon as possible.

**Comment by okv** [comment]
:ok_hand: will wait for it

**Comment by MikeDevice** [comment]
Made a lot of from [this comment](https://github.com/okv/east/pull/62#issuecomment-629640814). Now trying to fix tests and rest items from list.

File: lib/migrator/methods/_loadConfig.js [hunk 4]
```diff
@@ -35,11 +35,14 @@ module.exports = function _loadConfig(configPath) {
...
+				return {};
 			}
 		})
-		.then((configText) => {
-			const config = JSON.parse(configText);
+		.then((resultConfig) => {
+			const config = _.isString(resultConfig) ?
+				JSON.parse(resultConfig) :
+				resultConfig;
+
```
Comment (MikeDevice) [inline lib/migrator/methods/_loadConfig.js:45]: I've added that condition, because config may not necessarily be in json format.

**Comment by MikeDevice** [comment]
I faced an interesting error at tests:

```
Error: Cannot find module '../../adapter'
```

But if I change path here lib/migrator/methods/configure.js
```diff
// adapter path to require
- adapter: '../../adapter',
+ adapter: '../adapter',
```
tests don't crashes with that error. It's really interesting and mysterious error.

**Comment by MikeDevice** [comment]
Wow, that's really interesting. Now `loadModule` util's path is `lib/utils/loadModule.js`, but when I place it to the nested folder `lib/utils/loadModule/index.js` tests work fine. Interesting...

**Comment by MikeDevice** [comment]
So what do you think @okv?

**Comment by okv** [comment]
Hi, I have some thoughts about that errors but need to check it out. I'll try to do it soon.

**Comment by okv** [comment]
> I faced an interesting error at tests:
> 
> ```
> Error: Cannot find module '../../adapter'
> ```
> 
> But if I change path here `lib/migrator/methods/configure.js`
> 
> ```diff
> // adapter path to require
> - adapter: '../../adapter',
> + adapter: '../adapter',
> ```
> 
> tests don't crash with that error. It's really interesting and mysterious error.

I believe that error could be resolved by resolving path to default adapter at migrator configure - by doing something like `adapter: pathUtils.resolve(__dirname, '..', '..', adapter')` instead of `adapter: '../../adapter'`. @MikeDevice could try that?

**Comment by MikeDevice** [comment]
Yes, I'll do.

**Comment by MikeDevice** [comment]
Yes, that's working, but there is another problem. Inside `test/02-snapshot/bin/east/commands/init/withAlreadyExistingDirAndTraceFlag.js` snapshot should have absolute path to adapter, but in every environment it is different. I think about a fix.

**Comment by MikeDevice** [comment]
Wow, that's interesting. It seems like `testUtils/cleanSnapshotData.js` will help me.

**Comment by okv** [comment]
All paths for snapshots can be fixed at clear snapshot data function, you can find it at test utils.

On May 30, 2020 4:32:14 PM GMT+03:00, Mikhail Vyrodov <notifications@github.com> wrote:
>Yes, that's working, but there is another problem. Inside
>`test/02-snapshot/bin/east/commands/init/withAlreadyExistingDirAndTraceFlag.js`
>snapshot should have absolute path to adapter, but in every environment
>it is different. I think about a fix.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/okv/east/pull/62#issuecomment-636331481

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by okv** [comment]
Year, that function)

On May 30, 2020 4:37:59 PM GMT+03:00, Mikhail Vyrodov <notifications@github.com> wrote:
>Wow, that's interesting. It seems like `testUtils/cleanSnapshotData.js`
>will help me.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/okv/east/pull/62#issuecomment-636332180

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by MikeDevice** [comment]
I've made all items from them [list](https://github.com/okv/east/pull/62#issuecomment-629640814) and have fixed all of tests.

**Comment by MikeDevice** [comment]
By the way, I've noticed the ci checks passing take a long time.
1. Coverage script runs twice: for `version 14` and for `version stable`, because current stable version is 14.
2. Lint script runs for all versions.

There are tools for run scripts only for one ci process like [jest-ci](https://github.com/ai/jest-ci). May be it make sense to run coverage and lint like that lib?

**Comment by okv** [comment]
Great, I'll try to check it out shortly.

On May 30, 2020 6:17:32 PM GMT+03:00, Mikhail Vyrodov <notifications@github.com> wrote:
>I've made all items from them
>[list](https://github.com/okv/east/pull/62#issuecomment-629640814) and
>have fixed all of tests.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/okv/east/pull/62#issuecomment-636345202

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by okv** [comment]
Yes, it takes long time, but seems neither 1 nor 2 will help much. 1 - will work as expected (run tests against upcoming release) when nodejs 15 will arrive. 2 - might be done by shell conditions (as coverage currently) but linting is pretty fast for lib of such size. To significantly reduce (2x and more) testing time I think test runner strategy should be changed - it turns out that run every single test file in separate process costs a lot when you have many files ;) So I tend to return back and use mocha again (instead of tap). Maybe you have alternative ideas?

On May 30, 2020 6:25:27 PM GMT+03:00, Mikhail Vyrodov <notifications@github.com> wrote:
>By the way, I've noticed the ci checks passing take a long time.
>1. Coverage script runs twice: for `version 14` and for `version
>stable`, because current stable version is 14.
>2. Lint script runs for all versions.
>
>There are tools for run scripts only for one ci process like
>[jest-ci](https://github.com/ai/jest-ci). May be it make sense to run
>coverage and lint like that lib?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/okv/east/pull/62#issuecomment-636346201

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by MikeDevice** [comment]
> Maybe you have alternative ideas?

May be a custom test runner based on https://github.com/ai/jest-ci/blob/master/bin.js

```js
let ciJobNumber = require('ci-job-number')

if (ciJobNumber() === 1) {
  // run tests with coverage
} else {
  // run tests without coverage
}
```

File: lib/migrator/methods/_loadConfig.js (hunk 4 above)
Comment (okv) [reply lib/migrator/methods/_loadConfig.js:45]: There is a bug with config loading which introduced (apparently by me) in feature/v2 branch. I'll merge this PR and then will fix it properly.

**Comment by okv** [comment]
> > Maybe you have alternative ideas?
> 
> May be a custom test runner based on https://github.com/ai/jest-ci/blob/master/bin.js
> 
> ```js
> let ciJobNumber = require('ci-job-number')
> 
> if (ciJobNumber() === 1) {
>   // run tests with coverage
> } else {
>   // run tests without coverage
> }
> ```
> 
> or just use jest with this wrapper jest-ci.

Coverage already made only for node 14, anyway long tests are not big problem currently)

I've just reviewed the PR - nicely done :+1:  I'm going to merge it.
Thanks for your contribution.

**Pull request merged**